### ISSUE TITLE
secrets aren't preserving newlines encoded as strings within json fields

### ISSUE THREAD

**Issue by eleith** [woodpecker-ci/woodpecker#525]
### Component

server

### Describe the bug

reproducible steps

1. add a secret with the value `{"hi":"there\nshould be no new\nline control character"}`
2. in a pipeline, inject the secret as `VALUE` and run the command `echo $${VALUE}`
3. note that the newlines are now interpreted as control characters and aren't preserved in the string, if you were to save the value into a file, the file would look like the follow

 ```
{"hi":"there
should be no new
line control character"}
```

expected

1. the string would exactly match the format as entered `{"hi":"there\nshould be no new\nline"}`

### System Info

```shell
next-d02dfe99
```

### Additional context

```shell
the problem i'm trying to solve is to save a json file as a secret and in the pipeline, echoing that string back into a file.

i have tried manually adding the secret in the web-ui and using the cli (with the '@' curl syntax) and they all produced the same result.

i have verified in the database directly, the value is as expected (`{"hi":"there\nshould be no new\nline"}`)
```

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/woodpecker-ci/woodpecker/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://woodpecker-ci.org/docs/intro).
- [X] Check that there isn't [already an issue](https://github.com/woodpecker-ci/woodpecker/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A join our [Discord Chat Server](https://discord.gg/fcMQqSMXJy).

**Comment by anbraten** [comment]
I don't think we can easily support / solve this. Some users expect secrets to have proper new-lines (e.g. for ssh-keys). Replacing new-line with its escape version `\n` wouldn't work for them in that case I think.

**Comment by eleith** [comment]
to document my workaround, i currently take the secret, pipe it through sed and then store that in a file

```
echo $${SECRET} | sed -z 's/\n/\\\n/g' | sed -z 's/\\\n$//g' > /tmp/secret.json
```

now, any commands in my pipeline that need the secret can be pointed to a file, accessible to my pipeline

**Comment by eleith** [comment]
if JSON secrets can not be supported, it might be worth stating that in the [docs](https://woodpecker-ci.org/docs/usage/secrets#adding-secrets). 

it was unexpected for me to find that json was not being preserved as i saw it in the secrets UI, especially when i tried the curl syntax as suggested in the documentation:


> Loading secrets from file using curl @ syntax. This is the recommended approach for loading secrets from file to preserve newlines

**Comment by 6543** [comment]
well I think we can ajust things to make it work for everybody - but that will break things so v0.16.0 is appropriated :)

**Comment by 6543** [comment]
ok ... this is a bigger issue.

we do promote new lines down to the pipeline engines.

so it comes down to what **backend**&**shell** combination do support/convert newlines in env vars and/or how they are converted.

- local backend + fish = newlines and "\n" stay a simple string
- docker backend + fish = newlines become " " and "\n" stay a simple string
- local backend + zsh = newlines and "\n" are both interpreted as newline
- docker backend + bash = newlines become " " and "\n" stay a simple string
- ...

we as woodpecker can not do much as it even comes down to the linux api to create a command impl. in golang

so I suggest that we should add a hint to the secrets docs: if you have newlines, to preserve them the best way is to encode them base64

**Comment by xoxys** [comment]
I cant reproduce this behavior with the latest WP version. I have temp disabled, the secrets escaping in the logs:

```
+ echo ${SECRET} > /tmp/test.txt
+ cat /tmp/test.txt
{"hi":"there\nshould be no new\nline"}
+ echo ${SECRET}
{"hi":"there\nshould be no new\nline"}
```
